agent/pkg/cruntimes/crutils.go

```diff
@@ -0,0 +1,230 @@
+package cruntimes
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/url"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+
+	"github.com/determined-ai/determined/agent/pkg/docker"
+	"github.com/determined-ai/determined/agent/pkg/events"
+	"github.com/determined-ai/determined/master/pkg/archive"
+	"github.com/determined-ai/determined/master/pkg/cproto"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/syncx/waitgroupx"
+)
+
+var logLevel = regexp.MustCompile("(?P<level>INFO|WARN|ERROR|FATAL):    (?P<log>.*)")
+
+// ShipCmdLogs does what you might expect from its name.
+func ShipCmdLogs(
+	ctx context.Context,
+	r io.ReadCloser,
+	stdtype stdcopy.StdType,
+	p events.Publisher[docker.Event],
+) {
+	for scan := bufio.NewScanner(r); scan.Scan(); {
+		line := scan.Text()
+		if len(strings.TrimSpace(line)) == 0 {
+			continue
+		}
+
+		var level, log string
+		if matches := logLevel.FindStringSubmatch(line); len(matches) == 3 {
+			level, log = matches[1], matches[2]
+		} else {
+			level, log = model.LogLevelInfo, line
+		}
+
+		if err := p.Publish(ctx, docker.NewTypedLogEvent(level, log, stdtype)); err != nil {
+			logrus.WithError(err).Trace("log stream terminated")
+			return
+		}
+	}
+}
+
+// PprintCommand provides pretty printing of the given command to trace & log outputs.
+func PprintCommand(
+	ctx context.Context,
+	command string,
+	args []string,
+	p events.Publisher[docker.Event],
+	log *logrus.Entry,
+) error {
+	toPrint := command
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "--") { // print each arg on a new line
+			toPrint += " \\\n"
+			toPrint += "\t"
+			toPrint += arg
+		} else {
+			toPrint += " "
+			toPrint += arg
+		}
+	}
+
+	log.Trace(toPrint)
+	if err := p.Publish(ctx, docker.NewLogEvent(
+		model.LogLevelDebug,
+		toPrint,
+	)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CanonicalizeImage returns the canonicalized image name.
+func CanonicalizeImage(image string) string {
+	url, err := url.Parse(image)
+	isURIForm := err == nil
+	isFSForm := path.IsAbs(image)
```

Comment:
Yeah, I presume this is checking for a file on the file system as opposed to a docker reference.
If it is instead a docker reference (as typical) one needs to add docker:// for singularity.

I don't think we would ever add docker:// for podman though -- so unclear what this needs to do for podman.

Reply:
Made a ticket for this: [FOUNDENG-597](https://jira-pro.its.hpecorp.net:8443/browse/FOUNDENG-597)  Review use of singularity pull in the agent

Reply:
isURIForm is just checking if the image is specified as a URI (https://singularity-docs.readthedocs.io/en/latest/#supported-uris). isFSForm form just checks if it is an absolute path.

if it is an absolute path or it is a uri with a non empty scheme, we leave it alone.
otherwise assume it is a docker image and prepend `docker://`.

doc comment could probably say that better, my bad